utils/loading_cache.hh

```diff
@@ -391,11 +401,25 @@ private:
     /// Set the given item as the most recently used item.
     /// The MRU item is going to be at the front of the _lru_list, the LRU item - at the back.
     /// \param lru_entry Cache item that has been "touched"
-    void touch_lru_entry(ts_value_lru_entry& lru_entry) {
+    void touch_lru_entry_2_partitions(ts_value_lru_entry& lru_entry) {
         if (lru_entry.is_linked()) {
-            _lru_list.erase(_lru_list.iterator_to(lru_entry));
+            lru_list_type& lru_list = container_list(lru_entry);
+            lru_list.erase(lru_list.iterator_to(lru_entry));
+        }
+
+        if (lru_entry.touch_count() < PartitionHitThreshold) {
+            _logger.trace("Putting key {} into the new generation partition", lru_entry.key());
+            _new_gen_list.push_front(lru_entry);
+            lru_entry.inc_touch_count();
+        } else {
+            _logger.trace("Putting key {} into the old generation partition", lru_entry.key());
+            _lru_list.push_front(lru_entry);
+
+            // Bump it up only once to avoid a wrap around
+            if (lru_entry.touch_count() == PartitionHitThreshold) {
+                lru_entry.inc_touch_count();
```

Comment:
Unless I'm missing something, here instead of this if() and increment you can just do
```
lru_entry.touch_count = PartitionHitThreshold + 1
```
Or even PartitionHitThreshold (why the +1).

But, again, I think I'm misunderstanding something fundemental. I thought that the popular items will be expired (actually, demoted to regular LRU) by their popularity (LFU). If you want to do this, you need to remember the correct count - you can't just clip it at PartitionHitThreshold.

Again, it is possible I'm just misunderstanding the algorithm you're implementing :-(

Reply:
> Unless I'm missing something, here instead of this if() and increment you can just do ...

I don't want to do assignment every time because it would be the actual store.
Here I trade a store to a branch which is supposed to be well predicted in the good-case.

> I thought that the popular items will be expired (actually, demoted to regular LRU) by their popularity (LFU). If you want to do this, you need to remember the correct count - you can't just clip it at PartitionHitThreshold.
> Again, it is possible I'm just misunderstanding the algorithm you're implementing...

LRU is not implemented using this counter. Note that this counter has been added in this series but we had LRU here from day zero ;)

LRU is implemented using list - the recently touched items are moved to the front: https://github.com/scylladb/scylla/blob/master/utils/loading_cache.hh#L206

Reply:
Right. As I said above, I didn't understand that this (two ordinary LRU lists) is how this algorithm works. I didn't realize the popular items are also organized by LRU and not by popularity. Now I understand.